tiledb/common/thread_pool/thread_pool.h

```diff
@@ -47,7 +47,241 @@ namespace tiledb::common {
+    virtual Status get() = 0;
+  };
+
+  /**
+   * @brief Task class encapsulating std::future. Like std::future it's shared
+   * state can only be get once and thus only one thread. It can only be moved
+   * and not copied.
+   */
+  class Task : public ThreadPoolTask {
+   public:
+    /**
+     * Default constructor
+     * @brief Default constructed SharedTask is possible but not valid().
+     */
+    Task() = default;
+
+    /**
+     * Constructor from std::future
+     */
+    Task(std::future<Status>&& f, ThreadPool* tp)
+        : ThreadPoolTask(tp)
+        , f_(std::move(f)){};
+
+    /**
+     * Move constructor
+     */
+    Task(Task&& t) noexcept
+        : ThreadPoolTask(t.tp_)
+        , f_(std::move(t.f_)){};
+
+    /**
+     * Move assignenent from Task
+     */
+    Task& operator=(Task&& t) noexcept {
+      tp_ = t.tp_;
+      f_ = std::move(t.f_);
+      return *this;
+    }
+
+    /**
+     * Disable copy
+     */
+    Task(const Task&) = delete;
+
+    /**
+     * Disable copy assignment
+     */
+    Task& operator=(const Task&) = delete;
+
+    /**
+     * Wait in the threadpool for this task to be ready.
+     */
+    Status wait() {
+      if (tp_ == nullptr) {
+        throw std::runtime_error("Cannot wait, threadpool is not initialized.");
+      } else if (!f_.valid()) {
+        throw std::runtime_error("Cannot wait, task is invalid.");
+      } else {
+        return tp_->wait(this);
+      }
+    }
+
+    /**
+     * Is this task valid. Wait can only be called on vaid tasks.
+     */
+    bool valid() const noexcept {
+      return f_.valid();
+    }
+
+   private:
+    friend class ThreadPool;
+
+    /**
+     * Wait for input milliseconds for this task to be ready.
+     */
+    std::future_status wait_for(
+        const std::chrono::milliseconds timeout_duration) const {
+      return f_.wait_for(timeout_duration);
+    }
+
+    /**
+     * Get the result of that task. Can only be used once. Only accessible from
+     * within the threadpool `wait` loop.
+     */
+    Status get() {
+      return f_.get();
+    }
+
+    /**
+     * The encapsulates std::shared_future
+     */
+    std::future<Status> f_;
+  };
+
+  /**
+   * @brief SharedTask class encapsulating std::shared_future. Like
+   * std::shared_future multiple threads can wait/get on the shared state
+   * multiple times. It can be both moved and copied.
+   */
+  class SharedTask : public ThreadPoolTask {
+   public:
+    /**
+     * Default constructor
+     * @brief Default constructed SharedTask is possible but not valid().
+     */
+    SharedTask() = default;
+
+    /**
+     * Constructor from std::shared_future
+     */
+    SharedTask(std::shared_future<Status>&& f, ThreadPool* tp)
+        : ThreadPoolTask(tp)
+        , f_(std::move(f)){};
+
+    /**
+     * Constructor from std::future
+     */
+    SharedTask(std::future<Status>&& f, ThreadPool* tp) noexcept
+        : ThreadPoolTask(tp)
+        , f_(std::move(f)){};
```

Comment:
```suggestion
    /**
     * Constructor from std::future or std::shared_future
     */
    SharedTask(auto&& f, ThreadPool* tp)
        : ThreadPoolTask(tp)
        , f_(std::forward<decltype(f)>(f)){};
```

The constructors can be combined.